### ISSUE TITLE
Should I be getting ambiguity warnings?

### ISSUE THREAD

**Issue by matthew-dean** [langium/chevrotain-allstar#3]
Correct me if I'm wrong, but I would think that this algorithm would be tolerant with ambiguities in OR expressions. To be fair, this lookahead strategy eliminates most ambiguity warnings, but sometimes, some persist, and `IGNORE_AMBIGUITIES` has no effect. Is there a workaround?

**Comment by matthew-dean** [comment]
I had a look through the code, and AFAICT, `IGNORE_AMBIGUITIES` is never checked before building an ambiguity error?

**Comment by msujew** [comment]
You're right, we currently don't check the flag. Note that an ambiguity in this package hints at a very serious grammar flaw and should not be disregarded. Compared to the LL(k) lookahead, an ambiguity in ALL(*) leads to very serious performance degradation, as the algorithm doesn't stop until it is absolutely sure it's ambiguous. Likely what you're experiencing in https://github.com/langium/chevrotain-allstar/issues/5.

You can override the print behavior anyway with the constructor options of the strategy.

**Comment by msujew** [comment]
> I assume that gates are applied before checking for ambiguities?

Yes, similar to the default LL(k) behavior in Chevrotain, gates are checked beforehand. In ANTLR4, gates are checked on demand, but that also leads to performance issues in ambiguous grammars - just different ones. While chevrotain-allstar struggles if the gates are expensive, ANTLR4's strategy struggles in case of long prefixes (that could be resolved by performing the gate check beforehand). 

An adaptive gate evaluation mode (i.e. try lazy until encountering a long prefix - then switch to eager evaluation) would be quite interesting to investigate. Unfortunately nothing I have time for in the foreseeable future :(
